### PR TITLE
fix: white screen when invidious instances don't load

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
---
Fix: white screen when invidious instances don't load
---

**Important note**
We may remove your pull request if you do not use this provided PR template correctly.

**Pull Request Type**
Please select what type of pull request this is:
- [x] Bugfix
- [ ] Feature Implementation

**Related issue**
Please link the issue your pull request is referring to. If this pull request fully resolves the relevant issue, put "closes" before the issue number. Example: "closes #123456".

https://github.com/FreeTubeApp/FreeTube/issues/1689#issuecomment-926460495
Not sure if it closes the original issue but it fixes the issue mentioned in the comment

**Description**
File fallback & hardcoded fallback when api to retrieve instances fails

**Testing (for code that is not small enough to be easily understandable)**
Has this pull request been tested?
Please describe shortly how you tested it and whether there are any ramifications remaining. 
Yes, i changed the api link to an invalid link to test it

**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 0.14.0

### CODE REVIEW

File: src/renderer/store/modules/invidious.js [hunk 1]
```diff
@@ -21,25 +22,39 @@ const getters = {
...
+        return instance[1].uri.replace(/\/$/, '')
+      })
     } catch (err) {
       console.log(err)
-    }
-
-    const instances = response.filter((instance) => {
-      if (instance[0].includes('.onion') || instance[0].includes('.i2p')) {
-        return false
+      const fileName = 'invidious-instances.json'
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:43]:
```suggestion
      // Starts fallback strategy: read from static file
      // And fallback to hardcoded entry(s) if static file absent
      const fileName = 'invidious-instances.json'
```

File: src/renderer/store/modules/invidious.js [hunk 2]
```diff
@@ -21,25 +22,39 @@ const getters = {
...
+      let fileData
+      /* eslint-disable-next-line */
+      const fileLocation = payload.isDev ? './static/' : `${__dirname}/static/`
+      if (fs.existsSync(`${fileLocation}${fileName}`)) {
+        console.log('reading static file for invidious instances')
+        fileData = fs.readFileSync(`${fileLocation}${fileName}`)
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:49]: The new code for `instances` is specific to the expected content of the expected static file
```suggestion
        fileData = fs.readFileSync(`${fileLocation}${fileName}`)
        instances = JSON.parse(fileData).map((entry) => {
	        return entry.url
	    })
```

File: src/renderer/store/modules/invidious.js [hunk 3]
```diff
@@ -21,25 +22,39 @@ const getters = {
...
       } else {
-        return true
+        console.log('unable to read static file for invidious instances')
+        fileData = '[{"url": "https://invidious.snopyta.org"}]'
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:52]: `instances` is simply expected to store an array of URLs without any path
```suggestion
        instances = ['https://invidious.snopyta.org']
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

File: src/renderer/store/modules/invidious.js [hunk 4]
```diff
@@ -21,25 +22,41 @@ const getters = {
...
+      if (fs.existsSync(`${fileLocation}${fileName}`)) {
+        console.log('reading static file for invidious instances')
+        fileData = fs.readFileSync(`${fileLocation}${fileName}`)
+        instances = JSON.parse(fileData).map((entry) => {
+          return entry.url
+        })
       } else {
-        return true
+        console.log('unable to read static file for invidious instances')
+        instances = ["https://invidious.snopyta.org"]
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js]: I think my suggestion did use single quotes 😅 
(Though I like double quotes more :P)

File: src/renderer/store/modules/invidious.js [hunk 5]
```diff
@@ -21,25 +22,41 @@ const getters = {
...
       console.log(err)
-    }
-
-    const instances = response.filter((instance) => {
-      if (instance[0].includes('.onion') || instance[0].includes('.i2p')) {
-        return false
+      // Starts fallback strategy: read from static file
+      // And fallback to hardcoded entry(s) if static file absent
+      const fileName = 'invidious-instances.json'
+      let fileData
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:46]:
```suggestion
      const fileName = 'invidious-instances.json'
```

File: src/renderer/store/modules/invidious.js [hunk 6]
```diff
@@ -21,25 +22,41 @@ const getters = {
...
+      /* eslint-disable-next-line */
+      const fileLocation = payload.isDev ? './static/' : `${__dirname}/static/`
+      if (fs.existsSync(`${fileLocation}${fileName}`)) {
+        console.log('reading static file for invidious instances')
+        fileData = fs.readFileSync(`${fileLocation}${fileName}`)
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:51]:
```suggestion
        const fileData = fs.readFileSync(`${fileLocation}${fileName}`)
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]
Minor refactor

File: src/renderer/store/modules/invidious.js (hunk 4 above)
Comment (ChunkyProgrammer) [reply src/renderer/store/modules/invidious.js]: Yeah, that's my bad. I'm so used to using double quotes lol

**Review by PikachuEXE — APPROVED** [review]
You need 3 reviewers

**Comment by jasonhenriquez** [comment]
Were you able to test this, and/or is there a clear set of replication steps to induce this bug?

**Comment by ChunkyProgrammer** [comment]
> Were you able to test this, and/or is there a clear set of replication steps to induce this bug?

I tested it by setting the api link for fetching invidious instances to an invalid link

File: src/renderer/store/modules/invidious.js [hunk 7]
```diff
@@ -21,25 +22,40 @@ const getters = {
...
+      if (fs.existsSync(`${fileLocation}${fileName}`)) {
+        console.log('reading static file for invidious instances')
+        const fileData = fs.readFileSync(`${fileLocation}${fileName}`)
+        instances = JSON.parse(fileData).map((entry) => {
+          return entry.url
+        })
       } else {
-        return true
+        console.log('unable to read static file for invidious instances')
+        instances = ['https://invidious.snopyta.org']
```
Comment (peepopoggers) [inline src/renderer/store/modules/invidious.js:56]: you could add 'invidious.kavin.rocks' also as it is very reliable

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
havent tested yet but i would like to navigate the pages that dont require a internet connection. Such as about, settings, history, playlists and user profiles.

**Comment by PikachuEXE** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc 
I think that should be addressed in another PR

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Should i open an issue for that?

**Comment by PikachuEXE** [comment]
navigate the pages that dont require a internet connection?
ya

**Comment by peepopoggers** [comment]
@jasonhenriquez You are probably more knowledgable on some JS specifics than me, so you like the look of these changes?

**Comment by jasonhenriquez** [comment]
@peepopoggers I was having difficulty replicating this issue earlier & have not had much time to work on FreeTube recently, but I'll try again & report back to you.

File: src/renderer/store/modules/invidious.js [hunk 8]
```diff
@@ -21,25 +22,40 @@ const getters = {
...
+      if (fs.existsSync(`${fileLocation}${fileName}`)) {
+        console.log('reading static file for invidious instances')
+        const fileData = fs.readFileSync(`${fileLocation}${fileName}`)
+        instances = JSON.parse(fileData).map((entry) => {
+          return entry.url
+        })
       } else {
-        return true
+        console.log('unable to read static file for invidious instances')
+        instances = ['https://invidious.snopyta.org', 'https://invidious.kavin.rocks/']
```
Comment (PikachuEXE) [inline src/renderer/store/modules/invidious.js:56]:
```suggestion
        instances = [
          'https://invidious.snopyta.org', 
          'https://invidious.kavin.rocks/'
        ]
```

Off topic: I found it annoying that trailing commas are **disallowed**

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Review by PikachuEXE — APPROVED** [review]

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Im not sure how to properly test this

**Comment by ChunkyProgrammer** [comment]
> havent tested yet but i would like to navigate the pages that dont require a internet connection. Such as about, settings, history, playlists and user profiles.

I know this PR allows you to go to settings when internet is off so you could do that to test it

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
It didnt work for me :(

**Review by peepopoggers — APPROVED** [review]

**Comment by robotiko380** [comment]
Yes

________________________________________

**Comment by PikachuEXE** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc 

image

**Comment by peepopoggers** [comment]
yes

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Im a slowKING not poke

**Comment by PikachuEXE** [comment]
> Nvm it did work my machine was being a **slowpoke**

I am responding to this :)

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Ha forgot about that comment😅

**Pull request merged**